activerecord/lib/active_record/relation/where_clause.rb

```diff
@@ -81,7 +81,12 @@ def invert
       end
 
       def self.empty
-        new([], [])
+        @empty ||= begin
+          empty = new([], [])
+          # This method memoizes, which will blow up on a frozen object
+          empty.send(:referenced_columns)
+          empty.freeze
```

Comment:
What'd you think of overriding `#freeze` to memoize and super? Push this prepare-a-singleton duty out of this method.